### PR TITLE
Lh/basic UI

### PR DESCRIPTION
**Pull request by laholstege**

### CODE REVIEW

File: Download.Mvc/Client/components/FileDownload.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
+﻿import * as React from 'react';
+import { match } from 'react-router-dom';
```
Comment (srkirkland) [inline Download.Mvc/Client/components/FileDownload.tsx]: @laholstege doesn't seem like you need this match function

File: Download.Mvc/Client/components/FileDownload.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            .then(this.onLoad);
+    }
+    onLoad = () => {
+        window.location.href = this.state.link;
+    }
+
+    render() {
+        return (
+            <div>
+                {this.state.loading &&
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FileDownload.tsx:27]: I like this sytax.

File: Download.Mvc/Client/components/FileDownload.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            .then(link => this.setState({ link }))
+            .then(loading => this.setState({ loading: false }))
+            .then(this.onLoad);
+    }
+    onLoad = () => {
+        window.location.href = this.state.link;
+    }
+
+    render() {
+        return (
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FileDownload.tsx:25]: You could also have:
` if (!this.state.loading) return null; `

File: Download.Mvc/Client/components/FileDownload.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            link: '',
+            loading: true
+        };
+    }
+    componentDidMount = () => {
+        // grab link to download
+        fetch('/api/downloadFile/' + this.props.id, { credentials: 'same-origin' })
+            .then(response => response.json())
+            .then(link => this.setState({ link }))
+            .then(loading => this.setState({ loading: false }))
```
Comment (srkirkland) [inline Download.Mvc/Client/components/FileDownload.tsx]: @laholstege you should do this in one line, `.then(link => this.setState({ link, loading: false }))`

File: Download.Mvc/Client/components/FilePreview.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            link: '',
+        };
+    }
+    componentDidMount = () => {
+        // grab link to download
+        if (this.props.id) {
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FilePreview.tsx]: Pure opinion, but I prefer early exits:

`if (!this.props.id) return;`

File: Download.Mvc/Client/components/FileDownload.tsx (hunk 3 above)
Comment (srkirkland) [inline Download.Mvc/Client/components/FileDownload.tsx:25]: @laholstege this part is fine but another option would be to just say `if (!this.state.loading) return null;` and then do your regular return, no need for wrapping it in the this.state.loading && syntax

File: Download.Mvc/Client/components/FilePreview.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            fetch('/api/previewFile/' + this.props.id, { credentials: 'same-origin' })
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FilePreview.tsx]: Use string interpolation:

``` javascript
`/api/previewFile/${this.props.id}`
```

File: Download.Mvc/Client/components/FilePreview.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+                .then(response => response.json())
+                .then(link => this.setState({ link }));
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FilePreview.tsx]: You should add a .catch and log/display an error.

File: Download.Mvc/Client/components/FilePreview.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        }
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.id !== this.props.id) {
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FilePreview.tsx]: Early exit is particularly preferred for componentWillReceiveProps.

File: Download.Mvc/Client/components/FilePreview.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+            fetch('/api/previewFile/' + nextProps.id, { credentials: 'same-origin' })
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FilePreview.tsx]: String interpolation

File: Download.Mvc/Client/components/FileView.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    actions = [
+        { label: "Cancel", onClick: this.handleToggle },
+        { label: "Agree", onClick: this.downloadLink }
+    ];
+
+    componentDidMount = () => {
+        // go grab the file info we are looking at
+        fetch('/api/file/' + this.props.data.id, { credentials: 'same-origin' })
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FileView.tsx]: String interpolation

File: Download.Mvc/Client/components/FolderParent.tsx [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        super(props);
+    }
+    render() {
+        const iconStyle = {
+            'padding': '0px 5px 0px 0px'
+        };
+        return (
+            <div>
+                {
+                    this.props.data.parent && <div><i className="fa fa-folder" aria-hidden="true" style={iconStyle}></i>
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FolderParent.tsx]: Move your && into the bracket line as you've been doing elsewhere.

File: Download.Mvc/Client/components/FolderParent.tsx [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+                        <Link to={`../folder/${this.props.data.parent.id}`}>{this.props.data.parent.name}</Link></div>
+                }
+            </div>       
+            );
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FolderParent.tsx]: Whitespace

File: Download.Mvc/Client/components/FolderView.tsx [hunk 13]
```diff
@@ -14,23 +23,53 @@ export class FolderView extends React.Component<IProps, any> {
...
         // go grab the folder info we are looking at
         fetch('/api/folder/' + this.props.match.params.id, { credentials: 'same-origin' })
             .then(response => response.json())
             .then(data => this.setState({ data }));
     }
+
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.match.params.id !== this.props.match.params.id)
+        {
+            fetch('/api/folder/' + nextProps.match.params.id, { credentials: 'same-origin' })
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FolderView.tsx]: String interpolation

File: Download.Mvc/Client/components/FolderEntries.tsx [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        this.props.getReadmeID(id);
+    }
+
+    getLicenseID = (id) => {
+        this.props.getLicenseID(id);
+    }
+
+    render() {
+        let entryList = this.props.data.item_collection.entries.map((entry, index) =>
+            <FolderEntry key={index} entry={entry} updateReadmeID={this.getReadmeID} updateLicenseID={this.getLicenseID} licenseID={this.props.licenseID}/>
```
Comment (srkirkland) [inline Download.Mvc/Client/components/FolderEntries.tsx]: @laholstege you could always just call this.props.getReadmeID in here directly and save yourself the trouble of making local function, they don't actually do anything.

File: Download.Mvc/Client/components/FolderView.tsx [hunk 15]
```diff
@@ -1,6 +1,15 @@
...
 import 'isomorphic-fetch';
+import { FolderEntries } from './FolderEntries';
+
+export const listStyle = {
+    'listStyleType': 'none',
+    'padding': '0px 0px 0px 15px'
+};
+export const iconStyle = {
+    'padding': '0px 5px 0px 0px'
+};
```
Comment (jpknoll) [inline Download.Mvc/Client/components/FolderView.tsx]: Not sure how I feel about exporting styles...

File: Download.Mvc/Controllers/ApiController.cs [hunk 16]
```diff
@@ -23,7 +23,7 @@ public ApiController()
         public BoxClient Initialize()
         {
             var config = new BoxConfig("guako54vyy1hlzu2iblnrpki5npfr0oy", "zepso2qwwfySqdESmBtHBJcpfy9ajVnT", new Uri("http://localhost"));
-            var session = new OAuthSession("0pCQ9dWspzDWkEd2mFlHXusklzzLaZYk", "NOT_NEEDED", 3600, "bearer");
+            var session = new OAuthSession("hsmPGNdKJzC5r45Ecp4VsoaCFrFT6nny", "NOT_NEEDED", 3600, "bearer");
```
Comment (jpknoll) [inline Download.Mvc/Controllers/ApiController.cs:26]: Move this to an appsetting.

File: Download.Mvc/Controllers/ApiController.cs [hunk 17]
```diff
@@ -39,7 +39,7 @@ public BoxClient Initialize()
             const string topFolderId = "27707355823";
```
Comment (jpknoll) [inline Download.Mvc/Controllers/ApiController.cs:39]: Move this to an appsetting.

File: Download.Mvc/Client/components/FolderEntry.tsx [hunk 18]
```diff
@@ -0,0 +1,58 @@
+﻿import * as React from 'react';
+import { Link } from 'react-router-dom';
+import { iconStyle } from './FolderView';
+import { FileView } from './FileView';
+
+export class FolderEntry extends React.Component<any, any>{
+    constructor(props) {
+        super(props);
+        this.updateReadmeID = this.updateReadmeID.bind(this);
+        this.updateLicenseID = this.updateLicenseID.bind(this);
```
Comment (srkirkland) [inline Download.Mvc/Client/components/FolderEntry.tsx]: @laholstege if you just use the props directly, you can get rid of the local functions entirely.  If you want to keep them for some reason, if you change the signature to `updateReadmeID = (newID) => { ... }` then you don't need this bind in the constructor at all

**Review by jpknoll — CHANGES_REQUESTED** [review]

File: Download.Mvc/Client/components/FolderView.tsx [hunk 19]
```diff
@@ -14,23 +23,53 @@ export class FolderView extends React.Component<IProps, any> {
     constructor(props) {
         super(props);
 
+        this.getLicenseID = this.getLicenseID.bind(this);
+        this.getReadmeID = this.getReadmeID.bind(this);
```
Comment (srkirkland) [inline Download.Mvc/Client/components/FolderView.tsx]: @laholstege AFAIK you don't need these lines, ES6 style arrow functions are bound to the proper this scope by default

File: Download.Mvc/Client/components/FileView.tsx [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+        this.state = {
+            data: {},
+            active: false,
+            download: false
+        };
+    }
+
+    handleToggle = () => {
+        this.setState({ active: !this.state.active });
```
Comment (jSylvestre) [inline Download.Mvc/Client/components/FileView.tsx]: @laholstege You need to set download: false here as well.

**Review by jSylvestre — CHANGES_REQUESTED** [review]

**Comment by jSylvestre** [comment]
I'd change this to .. Otherwise when I'm looking at this it isn't clear that there may be other folders at the same level as xbox that I'm not seeing.

Either that or a more detailed tree view.

image

File: Download.Mvc/Client/components/FilePreview.tsx (hunk 7 above)
Comment (laholstege) [inline Download.Mvc/Client/components/FilePreview.tsx]: @jpknoll is there a reason you recommend this for Preview but not any other time I'm using fetch? 
also, catching looks like: 
```
      .catch((error) => {
        console.error(error);
      });
```
right?

**Pull request merged**